README.md
<!--
 * @Author: dongmin
 * @LastEditors: dongmin03
 * @Date: 2021-04-01 15:24:50
 * @LastEditTime: 2022-03-01 00:26:57
-->
# wangeditor-for-react

项目来源：[wangEditor](https://github.com/wangeditor-team/wangEditor/)

## 安装

```bash
npm install wangeditor-for-react react
```

## 使用

**事件监听**
```jsx
import ReactWEditor from 'wangeditor-for-react';

<ReactWEditor
  defaultValue={'<h1>标题</h1>'}
  linkImgCallback={(src,alt,href) => {
    // 插入网络图片的回调事件
    console.log('图片 src ', src)
    console.log('图片文字说明',alt)
    console.log('跳转链接',href)
  }}
  onlineVideoCallback={(video) => {
    // 插入网络视频的回调事件
    console.log('插入视频内容', video)
  }}
  onChange={(html) => {
    console.log('onChange html:', html)
  }}
  onBlur={(html) => {
    console.log('onBlur html:', html)
  }}
  onFocus={(html) => {
    console.log('onFocus html:', html)
  }}
/>
```

**自定义配置**

```jsx
import ReactWEditor from 'wangeditor-for-react';

<ReactWEditor
  placeholder="自定义 placeholder"
  config={{
    fontSizes: {
      'x-small': { name: '10px', value: '1' },
      small: { name: '12px', value: '2' },
      normal: { name: '16px', value: '3' },
      large: { name: '18px', value: '4' },
      'x-large': { name: '24px', value: '5' },
      'xx-large': { name: '32px', value: '6' },
      'xxx-large': { name: '48px', value: '7' },
    },
  }}
/>
```

**配置**

* [配置菜单](https://www.wangeditor.com/doc/pages/03-%E9%85%8D%E7%BD%AE%E8%8F%9C%E5%8D%95/)
* [内容校验](https://www.wangeditor.com/doc/pages/05-%E5%86%85%E5%AE%B9%E6%A0%A1%E9%AA%8C/)
* [粘贴过滤](https://www.wangeditor.com/doc/pages/06-%E7%B2%98%E8%B4%B4%E8%BF%87%E6%BB%A4/)
* [上传图片](https://www.wangeditor.com/doc/pages/07-%E4%B8%8A%E4%BC%A0%E5%9B%BE%E7%89%87/)
* [上传视频](https://www.wangeditor.com/doc/pages/07-%E4%B8%8A%E4%BC%A0%E8%A7%86%E9%A2%91/)

**使用 Ref**

通过 `Ref` 获取所有 API

```jsx
import React, { useRef } from 'react';
import ReactWEditor from 'wangeditor-for-react';

function App() {
  let editorRef = useRef(null)
  return (
    <ReactWEditor
      ref={editorRef}
      onBlur={(html) => {
        if (editorRef.current) {
          console.log('ref', editorRef.current.editor.txt.append('追加内容'))
        }
      }}
    />
  );
}

export default App;
```

[常用 API](https://www.wangeditor.com/doc/pages/08-%E5%B8%B8%E7%94%A8API/)

**销毁编辑器**
```jsx
import React, { useRef } from 'react';
import ReactWEditor from 'wangeditor-for-react';

function App() {
  let editorRef = useRef(null)
  return (
    <>
    <ReactWEditor ref={editorRef} />
    <a href="#" onClick={() => editorRef.current.destroy()}>click</a>
    </>
  );
}

export default App;
```

**多语言**

首先需要安装语言包
```bash
npm i -S i18next
```

使用
```jsx
import { extend } from 'wangeditor-for-react';
import i18next from 'i18next';

const ReactWEditorOfLang = extend({ i18next })

<ReactWEditorOfLang
  config={{
    lang: 'en',
  }}
/>
```

自定义语言

```jsx
import { extend } from 'wangeditor-for-react';
import i18next from 'i18next';

const ReactWEditorOfLang = extend({ i18next })

<ReactWEditorOfLang
  config={{
    lang: 'japan',
  }}
  languages={{
    japan: {
      wangEditor: {
        请输入正文: '本文を入力してください',
      },
    }
  }}
/>
```
## 增加 hook
分为全局钩子 `globalHook` 和实例钩子 `instanceHook`。

`globalHook` 注册在构造函数上，即 `new E()` 中的 `E`。`instanceHook` 注册在实例后，`editor.create()` 执行前。

钩子支持分割符，例如 `menus.extend` 相当于 `this.editor.menus.extend`。

钩子有两种值，一种是数组，一种是函数。

```jsx
<ReactWEditor
  instanceHook={{
    // 使用数组时，通常 key 代表的钩子是一个方法，此处 menus.extend 是个方法，那么数组就是其参数。
    'menus.extend': ['alertMenuKey', AlertMenu],
    // 使用方法是，通常 key 代表的钩子是一个对象，可以利用方法来绑定。方法的形参第一位是当前实例的 editor，后面依次是 key 分割代表的对象。
    'config.menus': function(editor, config, menus) {
      config.menus = menus.concat("alertMenuKey")
    }
  }}
/>
```


使用 `globalHook` 扩展菜单，`AlertMenu` 参考 [Button 菜单](https://www.wangeditor.com/doc/pages/11-%E8%87%AA%E5%AE%9A%E4%B9%89%E6%89%A9%E5%B1%95%E8%8F%9C%E5%8D%95/02-Button%E8%8F%9C%E5%8D%95.html)。
```jsx
import React from 'react';
import ReactWEditor from 'wangeditor-for-react';
import AlertMenu from './AlertMenu';

function App() {
  return (
    <ReactWEditor
      globalHook={{
         registerMenu: ['alertMenuKey', AlertMenu]
      }}
    />
  );
}

export default App;
```

使用 `instanceHook` 扩展菜单。
```jsx
import React from 'react';
import ReactWEditor from 'wangeditor-for-react';
import AlertMenu from './AlertMenu';

function App() {
  return (
    <ReactWEditor
      instanceHook={{
        'menus.extend': ['alertMenuKey', AlertMenu],
        'config.menus': function(editor, config, menus) {
          config.menus = menus.concat("alertMenuKey")
        }
      }}
    />
  );
}

export default App;
```


## TODO

* 支持 [scroll-to-head](https://www.wangeditor.com/doc/pages/08-%E5%B8%B8%E7%94%A8API/06-scroll-to-head.html)
* 支持 [菜单和编辑区域分离](https://www.wangeditor.com/doc/pages/01-%E5%BC%80%E5%A7%8B%E4%BD%BF%E7%94%A8/03-%E8%8F%9C%E5%8D%95%E5%92%8C%E7%BC%96%E8%BE%91%E5%8C%BA%E5%9F%9F%E5%88%86%E7%A6%BB.html)

CHANGELOG.md
<!--
 * @Author: donggg
 * @LastEditors: dongmin03
 * @Date: 2021-04-27 19:13:31
 * @LastEditTime: 2022-03-07 17:36:51
-->
## 1.5.6 - 2022-03-07
- chore: 更新 wangeditor 版本
- fix: 包的完整
- chore: react 依赖调整 (#23)

## 1.5.3 - 2021-08-23
- fix: 修复类型应用问题

## 1.5.2 - 2021-08-20
- fix: 修复类型定义问题
  
## 1.5.1 - 2021-08-08
- chore: 增加类型声明文件

## 1.5.0 - 2021-07-05
- chore: 删除 babel 编译及相关依赖
- refactor: 重构为 ts
- feat: 升级 wangeditor 版本
- doc: 修改官方文档最新地址
- chore: 增加 lint 规范

## 1.4.1 - 2021-06-25
- fix: 修复受控组件强制替换 html 内容
- chore: 修改脚本

## 1.4.0 - 2021-06-05
- feat: 增加钩子方法

## 1.3.0 - 2021-05-15
- perf: 更新 wangeditor 版本
- chore: 自动发布脚本
- 更新 CHANGELOG

## 1.2.2 - 2021/05/11
- feat: 多语言支持
- feat: 增加销毁方法
- feat: 支持 onlineVideoCallback 和 linkImgCallback 回调函数
- doc: 修改了介绍

## 1.1.3 - 2021/05/08
- fix: 兼容 v3 版本
- feat: 容器支持内联样式与样式类
- feat: 追加 onFous 与 onBlur 事件

## 1.1.2 - 2021/05/06
- fix: 设置 localBlobImg 修复无图片时的错误提醒

## 1.0.1 - 2021/04/27
- feat: 增加 Blob 本地图片，并在类中提供处理函数 replaceHTMLImgBlobURL

```jsx
import React from 'react';
import ReactWEditor from 'wangeditor-for-react';

class Demo extends React.Component {
  submit () {
    if (this.editorRef) {
      // 当前曾经添加过所有的图片文件，包括在编辑器中删除的图片，key 是 blob 地址，value 是 File
      const imgs = this.editorRef.imgFile.getAllImgFiles()
      console.log('imgs', imgs)

      // 根据 blob 地址，替换 html 中图片地址
      const content = this.editorRef.replaceHTMLImgBlobURL(
        '<img src="blob:http://localhost/330f3a67-7be9-4871-8f94-473c45c1f524>',
        file =>
          `/image?filename=${file.lastModified}_${file.name}`,
      )
      console.log('content', content)
    }
  }

  render () {
    return (
    <div>
      <ReactWEditor
        localBlobImg
        ref={ref => this.editorRef = ref}
      />
      <a onClick={this.submit}>submit</a>
    </div>
    )
  }
}
```


es6/core.js
/*
 * @Author: dongmin
 * @LastEditors: donggg
 * @Date: 2021-04-01 19:02:53
 * @LastEditTime: 2021-06-25 18:36:17
 */
import React from 'react';
import WEdtior from 'wangeditor';
import createId from './utils/unique-id';
import { isObject, isString } from './utils/isType';
import { isEqualString, isEmpty, difference } from './utils/helper';
import { replaceHTMLImgBlobURL } from './utils/htmlRnder';
import ImgFile from './imgFile'


export default class ReactWEditor extends React.PureComponent {
  // #Private
  id = createId(8);
  // #Private
  editor = null;
  // #Private
  hasCreated = false; // 是否执行 create 创建函数
  // #Protected
  defaultConfig = {
    zIndex: 1,
  };

  imgFile = new ImgFile()

  componentDidMount() {
    try {
      this.init();
      this.create();
    } catch (e) {
      console.error(`[ReactWEdtior Error]: ${e}`)
    }
  }

  componentWillUnmount() {
    if (this.editor) {
      this.editor.destroy()
    }
  }

  // #Private
  __hook__run = (hooks = [], args = [], target = {}) => {
    hooks.forEach((hook, index) => {
      if (!hook) {
        return;
      }
      
      if (target[hook] && typeof target[hook] === 'function' && args[index]) {
        target[hook].apply(target[hook], args[index])
      } else if (/^(\w+\.\w+)+$/.test(hook) && args[index]) {
        const path = hook.split('.')
        const cache = [];
        let fn = target;
        path.forEach((d) => {
          cache.push(fn);
          fn = fn[d]
        })
        cache.push(fn);
        if (typeof fn === 'function') {
          fn.apply(cache[cache.length-2], args[index])
        } else if(typeof args[index] === 'function') {
          args[index].apply(args[index], cache)
        }
      }
    })
  }

  // #Private
  __before__instanced() {
    const { globalHook = {} } = this.props;
    const hooks = Object.keys(globalHook);
    const args = Object.values(globalHook);

    this.__hook__run(hooks, args, WEdtior);
  }

  // #Private
  __after__instanced() {
    if (!this.check()) {
      return;
    }
    const { instanceHook = {} } = this.props;
    const hooks = Object.keys(instanceHook);
    const args = Object.values(instanceHook);

    this.__hook__run(hooks, args, this.editor);
  }

  // #Protect
  init() {
    const elem = document.getElementById(`editor-${this.id}`);
    if (elem) {
      // 0. 初始化前，调用全局的 hook
      this.__before__instanced()

      // 1. 初始化
      this.editor = new WEdtior(`#editor-${this.id}`);

      // 2. 初始化后，调用实例的 hook，支持相对路径，例如键值是 'menus.extend'
      this.__after__instanced()

      // 3. 根据属性配置默认设置
      this.setDefaultConfigByProps();

      // 4. 根据默认设置更新设置
      this.setConfig(this.defaultConfig);
    } else {
      console.error('[ReactWEdtior Error]: dom is not found')
    }
  }

  // #Protect
  check() {
    if (this.editor) {
      return true;
    }
    console.error('[ReactWEdtior Error]: editor not found')
    return false;
  }

  // #Protect
  create(context = {}) {
    const { config, defaultValue, customConfig } = this.props;
    if (this.check()) {
      // 1. 根据 config 属性配置设置
      this.setConfig(config || customConfig)

      // 2. 扩展 edtior
      this.extend(context)

      // 3. 生成 editor
      this.editor.create();

      // 4. 修改标识
      this.created()

      // 5. 根据 defaultValue 设置内容
      this.setContentByHTMLString(defaultValue)
    }
  }
  
  /**
   * 通过 context 扩展 edtior 
   * @param {object} context 待扩展的内容
   * @param {array} customFilter 需要过滤的扩展字段
   */
   extend(
    context = {}, 
    customFilter = []
  ) {
    if (this.check()) {
      // 1. 过滤数组
      const filter = Object.keys(this.editor).concat(customFilter || [])

      // 2. 向 editor 上扩展
      difference(Object.keys(context), filter).forEach((key) => this.editor[key] = context[key])
    }
  }

  /**
   * 销毁编辑器
   */
  destroy() {
    if (!this.isCreated()) {
      console.error('[ReactWEdtior Error]: editor has not created, don\'t destroy.')
      return;
    }
    // 1. 销毁
    this.editor.destroy();
    this.editor = null;

    // 2. 修改标识
    this.destroyed()
  }

  /**
   * 配置 editor
   * @param {*} config 配置
   * @doc https://www.wangeditor.com/doc/
   */
  setConfig(config) {
    if (config && isObject(config)) {
      this.editor.config = Object.assign(this.editor.config, config);
    }

    // 多语言处理
    const { languages } = this.props;
    if (languages && isObject(languages) &&!isEmpty(languages)) {
    	this.editor.config.languages = Object.assign(this.editor.config.languages, languages)
    }
  }

  /**
   * 根据属性，配置默认设置
   */
  setDefaultConfigByProps = () => {
    const {
      placeholder,
      onChange,
      onFocus,
      onBlur,
      linkImgCallback,
      onlineVideoCallback,
      localBlobImg,
    } = this.props;

    if (placeholder) this.defaultConfig.placeholder = placeholder
    if (onChange) this.defaultConfig.onchange = onChange
    if (onFocus) this.defaultConfig.onfocus = onFocus
    if (onBlur) this.defaultConfig.onblur = onBlur
    if (linkImgCallback) this.defaultConfig.linkImgCallback = linkImgCallback
    if (onlineVideoCallback) this.defaultConfig.onlineVideoCallback = onlineVideoCallback

    // 图片替换为本地Blob伪URL
    if (localBlobImg) {
      this.defaultConfig.customUploadImg = (resultFiles, insertImgFn) => {
        resultFiles.forEach((file) => {
          const url = URL.createObjectURL(file)
          this.imgFile.saveImgFiles(url, file)
          insertImgFn(url)
        })
      }
    }
  }

  /**
   * 设置 editor 内容。注意！必须在创建完 editor 后才可以设置内容
   * @param {string} html 回填的 html 字符串
   */
  setContentByHTMLString(html) {
    if (!this.isCreated()) {
      console.error('[ReactWEdtior Error]: editor has not created')
    }

    if (this.check() && isString(html)) {
      try {
        this.editor.txt.html(html)
      } catch (e) {
        console.error(`[ReactWEdtior Error]: ${e}`)
      }
    }
  }

  /**
   * 替换 html 中的 img 标签的 src 引用地址
   * @param {string} html html 文本
   * @param {function} callback 替换过程中的回调函数
   * @returns 替换后的 html 文本
   */
  replaceHTMLImgBlobURL(html, callback) {
    return replaceHTMLImgBlobURL(html, this.imgFile.getAllImgFiles(), callback)
  }

  render() {
    const { style, className } = this.props;
    return <div style={style} className={className} id={`editor-${this.id}`} />
  }

  changeCreatedFlag = flag => this.hasCreated = flag;
  created = () => this.changeCreatedFlag(true);
  destroyed = () => this.changeCreatedFlag(false);
  isCreated = () => this.hasCreated === true
}
